torch/utils/checkpoint.py

```diff
@@ -21,38 +21,52 @@ def check_backward_validity(inputs):
         warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
 
 
-# Global switch to toggle whether or not checkpointed passes stash and restore
-# the RNG state.  If True, any checkpoints making use of RNG should achieve deterministic
-# output compared to non-checkpointed passes.
-preserve_rng_state = True
+# We can't know if the run_fn will internally move some args to a device
+# other than the current device, which would require logic to preserve
+# rng states for those devices as well.  We could paranoically stash and restore
+# ALL the rng states for all visible devices, but that seems very wasteful for
+# most cases.  Compromise:  Stash the RNG state for the current device as well as
```

Comment:
Nooo, there's no point in stashing the values on the current device I think. What this compromise is, is an assumption that you will only do compute on the devices of your inputs. We should probably just stash for all devices you can find among them, and only those

Reply:
Ok, changed in latest commit